core/Test/Tasty/Run.hs

```diff
@@ -257,89 +252,131 @@ instance Show DependencyException where
+  -- to run them, and meta information needed to watch test progress and calculate
+  -- dependencies in 'resolveDeps'.
+  unresolvedTestTree <-
+    flip runReaderT mempty $
+      foldTestTree0 (pure (TGroup [])) (TreeFold { .. }) opts0 tree
+
   let
-    traversal :: Tr
-    traversal =
-      foldTestTree
-        (trivialFold :: TreeFold Tr)
-          { foldSingle = runSingleTest
-          , foldResource = addInitAndRelease
-          , foldGroup = \_opts name (Traversal a) ->
-              Traversal $ mapWriterT (local (first (Seq.|> name))) a
-          , foldAfter = \_opts deptype pat (Traversal a) ->
-              Traversal $ mapWriterT (local (second ((deptype, pat) :))) a
-          }
-        opts0 tree
-  (tests, fins) <- unwrap (mempty :: Path) (mempty :: Deps) traversal
-  let
-    mb_tests :: Either [[Path]] [(Action, TVar Status)]
-    mb_tests = resolveDeps $ map
-      (\(act, testInfo) ->
-        (act (Seq.empty, Seq.empty), testInfo))
-      tests
-  case mb_tests of
-    Right tests' -> return (tests', fins)
-    Left cycles -> throwIO (DependencyLoop cycles)
+    finalizers = collectFinalizers unresolvedTestTree
+    tests = collectTests (resolveTestActions unresolvedTestTree)
+
+  case resolveDeps tests of
+    Right tests' -> return (tests', finalizers)
+    Left cycles  -> throwIO (DependencyLoop cycles)
 
   where
-    runSingleTest :: IsTest t => OptionSet -> TestName -> t -> Tr
-    runSingleTest opts name test = Traversal $ do
-      statusVar <- liftIO $ atomically $ newTVar NotStarted
-      (parentPath, deps) <- lift ask
+    -- * Functions used in 'TreeFold'
+    foldSingle :: IsTest t => OptionSet -> TestName -> t -> Tr
+    foldSingle opts name test = do
+      testStatus <- liftIO $ newTVarIO NotStarted
+      (parentPath, testDeps) <- ask
       let
-        path = parentPath Seq.|> name
-        act (inits, fins) =
-          executeTest (run opts test) statusVar (lookupOption opts) inits fins
-      tell ([(act, (statusVar, path, deps))], mempty)
-    addInitAndRelease :: OptionSet -> ResourceSpec a -> (IO a -> Tr) -> Tr
-    addInitAndRelease _opts (ResourceSpec doInit doRelease) a = wrap $ \path deps -> do
-      initVar <- atomically $ newTVar NotCreated
-      (tests, fins) <- unwrap path deps $ a (getResource initVar)
-      let ntests = length tests
-      finishVar <- atomically $ newTVar ntests
+        testPath = parentPath |> name
+        testAction = executeTest (run opts test) testStatus (lookupOption opts)
+      pure $ TAction (TestAction {..})
+
+    foldResource :: OptionSet -> ResourceSpec a -> (IO a -> Tr) -> Tr
+    foldResource _opts (ResourceSpec doInit doRelease) a = do
+      initVar <- liftIO $ newTVarIO NotCreated
+      testTree <- a (getResource initVar)
+      let ntests = length (collectTests testTree) -- TODO: Store size of collection
+      finishVar <- liftIO $ newTVarIO ntests      --       in data type?
```

Comment:
Any resolution on TODO?

Reply:
Picked up in `Cache tree size in 'TestActionTree'`. It is slightly unsafe, see documentation of `TGroup`. I tried moving it to its own module doing the safe constructor thingy, but it dragged in a whole slush of types defined in `Run.hs`. That didn't really help readability IMO, so I left it as it currently is.